shepherd/LCM.py

```diff
@@ -1,41 +1,169 @@
 import threading
 import json
-import lcm # pylint: disable=import-error
+import socket
+import threading
+import selectors
+
+SERVER_ADDR = ('127.0.0.1', 5001) # doesn't need to be available on network
+
+
+"""
+Message format between client and server is:
+4 byte int, little endian: len1
+4 byte int, little endian: len2
+len1 bytes, utf-8 string
+len2 bytes, utf-8 string
 
-LCM_address = 'udpm://239.255.76.68:7667?ttl=2'
+Normally first and second string represent a target+message.
+(target, "") from client to server means subscribe to target
+"""
+
+
+CLIENT_THREAD = None
 
 def lcm_start_read(receive_channel, queue, put_json=False):
     '''
     Takes in receiving channel name (string), queue (Python queue object).
     Takes whether to add received items to queue as JSON or Python dict.
     Creates thread that receives any message to receiving channel and adds
-    it to queue as tuple (header, dict).
+    it to queue as tuple (header, dict). Multiple calls to lcm_start_read
+    with the same target will replace the old queue.
     header: string
     dict: Python dictionary
     '''
-    comm = lcm.LCM(LCM_address)
+    start_client_thread_if_not_alive()
+    if receive_channel not in CLIENT_THREAD.open_targets:
+        send_message(CLIENT_THREAD.conn, receive_channel, "")
+    CLIENT_THREAD.open_targets[receive_channel] = (queue, put_json)
+
+def lcm_send(target_channel, header, dic=None):
+    '''
+    Send header and dictionary to target channel (string)
+    '''
+    start_client_thread_if_not_alive()
+    if dic is None:
+        dic = {}
+    dic['header'] = header
+    json_str = json.dumps(dic)
+    send_message(CLIENT_THREAD.conn, target_channel, json_str)
+
+def start_client_thread_if_not_alive():
+    global CLIENT_THREAD
+    if CLIENT_THREAD is None:
+        CLIENT_THREAD = ClientThread()
+        CLIENT_THREAD.start()
+
+class ClientThread(threading.Thread):
+    def __init__(self):
+        super().__init__()
+        self.daemon = True #will be shut down abruptly when main thread dies
+        self.conn = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.selobj = SelObject()
+        self.open_targets = {}
+        self.conn.connect(SERVER_ADDR)
+    
+    def run(self):
+        while True:
+            data = self.conn.recv(1024)
+            if len(data) == 0:
+                break
+            else:
+                self.selobj.inb += data;
+                for target, message in self.selobj:
+                    self.forward_message(target, message)
+        self.conn.close()
 
-    def handler(_, item):
+    def forward_message(self, target, message):
+        queue, put_json = self.open_targets[target]
```

Comment:
Does this error when messages are sent to a target that this thread did not subscribe to?

Reply:
LCM backend should only forward messages that this client subscribed to